File: 03/sol.py
import re


def main():
    mul_regex = re.compile(r"mul\((\d+),(\d+)\)", re.MULTILINE)
    mul_do_regex = re.compile(
        r"(?:mul\((\d+),(\d+)\)|(do)\(\)|(don't)\(\))", re.MULTILINE
    )
    instructions = []
    instructions2 = []
    sum = 0
    sum2 = 0
    with open("in.txt", "r") as f:
        lines = f.readlines()
        for line in lines:
            res = mul_regex.findall(line)
            for i in res:
                instructions.append(i)

        for line in lines:
            res = mul_do_regex.findall(line)
            for i in res:
                instructions2.append(i)

    # print(instructions)
    for instruction in instructions:
        sum += int(instruction[0]) * int(instruction[1])

    print(sum)

    isDo = True
    for instruction in instructions2:
        if "do" in instruction:
            isDo = True
        elif "don't" in instruction:
            isDo = False

        if isDo:
            numbers = [x for x in instruction if x.isdigit()]

            sum2 += int(numbers[0]) * int(numbers[1]) if len(numbers) == 2 else 0

    print(sum2)


if __name__ == "__main__":
    main()

File: 03/sol.exs
defmodule Solution do
  def main do
    mul_regex = ~r/mul\((\d+),(\d+)\)/
    mul_do_regex = ~r/(?:mul\((\d+),(\d+)\)|(do)\(\)|(don't)\(\))/

    lines = File.read!("in.txt") |> String.split("\n")

    # Part 1
    sum =
      lines
      |> Enum.flat_map(fn line ->
        Regex.scan(mul_regex, line, capture: :all_but_first)
      end)
      |> Enum.reduce(0, fn [num1, num2], acc ->
        acc + String.to_integer(num1) * String.to_integer(num2)
      end)

    IO.puts(sum)

    # Part 2
    instructions =
      lines
      |> Enum.flat_map(fn line ->
        Regex.scan(mul_do_regex, line)
      end)

    {sum2, _} =
      instructions
      |> Enum.reduce({0, true}, fn [_match | captures], {acc, is_do} ->
        cond do
          Enum.any?(captures, &(&1 == "do")) ->
            {acc, true}

          Enum.any?(captures, &(&1 == "don't")) ->
            {acc, false}

          is_do ->
            numbers =
              captures
              |> Enum.filter(&(not is_nil(&1)))
              |> Enum.filter(&Regex.match?(~r/^\d+$/, &1))

            case numbers do
              [num1, num2] -> {acc + String.to_integer(num1) * String.to_integer(num2), is_do}
              _ -> {acc, is_do}
            end

          true ->
            {acc, is_do}
        end
      end)

    IO.puts(sum2)
  end
end

Solution.main()
